### PR TITLE
feat: add support for Jira Data Center custom fields via config

### PR DESCRIPTION
**Pull request by SalimGrayaa**
# Pull Request

## Summary

add a feature where user can add custom fields to the metadata. 

## Type of change

- [x] Feature
- [ ] Bug fix
- [ ] Docs update
- [ ] Chore

## Docs Impact (required for any code or docs changes)

- [x] Does this change require docs updates? If yes, list pages:
  - docs\users\detailed-guides\data-sources\jira.md
  - docs\users\configuration\config-file-reference.md
- [x] Have you updated or added pages under `docs/`?
- [x] Did you build the site and run the link checker? (`python website/build.py` + `python website/check_links.py`)
- [ ] Did you avoid banned placeholders? (no "TBD/coming soon")

## Testing

This is the configuration file:
projects:
  my-project:
    sources:
      jira:
        projectkey:
          base_url: https://jira.softathome.com/
          project_key: projectkey
          deployment_type: "datacenter"
          token: ${JIRA_TOKEN}
          enable_file_conversion: true
          extra_fields:
          - param_name: customfield_12000
            name: detected_in_version
            field_type: simple
          - param_name: versions
            name: affected_versions
            field_type: array_object
            attr_name: name
Then I executed the following command: uv run qdrant-loader ingest --workspace . --log-level debug 
Using debug mode, I verified that the fields exist in the logs. The fields are also successfully added to Qdrant.
 Output:
image

## Checklist

- [x] Tests pass (`pytest -v`)
- [x] Linting passes (make lint / make format)
- [x] Documentation updated (if applicable)

---

> [!NOTE]
> **Medium Risk**
> Adds dynamic field extraction and injects values into `JiraIssue`/document metadata; misconfiguration or missing propagation (e.g., connectors not passing `extra_fields`) could cause runtime errors or unexpected metadata shapes.
> 
> **Overview**
> Adds configurable `extra_fields` support for the Jira connector, letting users specify additional Jira field parameters and how to extract them (simple/object/array/array-of-objects) via new `JiraExtraField`/`JiraFieldType` config models with validation.
> 
> Issue parsing now optionally extracts these fields and attaches them onto `JiraIssue` (model now allows extra attributes), and document generation appends the extracted values into `Document.metadata`. The Data Center connector is updated to pass configured `extra_fields` into issue parsing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 46b39c705139c5cc76d6a655affaa1fb502d6c1a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Configure and extract additional Jira fields into document metadata (simple, nested object, and array-of-object types).

* **Improvements**
  * Issue parsing now includes configured extra fields so their values appear in produced documents.
  * Embedding responses are aligned to original inputs; empty embeddings are skipped and logged.

* **Validation**
  * Extra-field definitions are trimmed, non-empty, enforce attribute mapping for nested types, and disallow duplicate names.

* **Bug Fixes**
  * Clearer error handling when project state update conflicts occur.

* **Style**
  * Minor footer and layout HTML tweaks.

[Review Change Stack](https://app.coderabbit.ai/change-stack/martin-papy/qdrant-loader/pull/251)

### CODE REVIEW

**Comment by thanhbuingoc-cbtw** [comment]
Hi @SalimGrayaa, this is a really good idea and honestly a much-needed feature — custom fields are pretty common on Data Center Jira setups, so thanks for working on this!

  That said, before merging I think we'd want a couple more things:

  1. Docs — could you add extra_fields to docs/users/detailed-guides/data-sources/jira.md? Somewhere under "Advanced Configuration" or "Configuration Options" feels natural. A short YAML snippet showing each
  of the four field_type values (simple, object, array, array_object) would go a long way in helping users pick the right one.

  2. Tests — I couldn't find any tests covering the new code. A few cases I'd love to see:

  - JiraExtraField validation: attr_name required when field_type is object / array_object, and name rejected when it collides with a reserved attribute
  - validate_extra_fields_unique: duplicate param_name or name raises
  - _extract_extra_field_value: happy path for each of the 4 types, plus edge cases (missing param_name in response, non-dict item in array_object)
  - End-to-end via get_documents: extra fields actually show up in Document.metadata

  test_get_cloud_documents in tests/unit/connectors/jira/test_jira_connector.py is a nice template — you can just extend the fixture with a custom field.

  Thanks again!

**Comment by SalimGrayaa** [comment]
Hi @thanhbuingoc-cbtw , I’ve added the requested modifications:

1. Updated the documentation at docs/users/detailed-guides/data-sources/jira.md under the "Advanced Configuration" section.

2. Added test_extra_fields.py to cover:
    - JiraExtraField validation
    - validate_extra_fields_unique
    - _extract_extra_field_value
image

3. Added four end-to-end tests via get_documents in tests/unit/connectors/jira/test_jira_connector.py.
image


Please let me know if any further updates are needed.

**Comment by minh362000** [comment]
Hi @SalimGrayaa, thank you again for your contribution and support.
At the moment, our team is still refining the process for merging external contributions into the development/release branches, including changelog updates and release planning. We need a bit more time to align everything internally.

We’ll update you once the process is finalized and let you know about the next steps for this PR.

**Comment by SalimGrayaa** [comment]
Hi @minh362000 ,
Thank you for the update and for keeping me informed. I completely understand the need to refine the contribution and release process internally. I’m happy to wait for the next steps regarding the PR. Looking forward to hearing from you once everything is finalized.

**Comment by minh362000** [comment]
Hi @SalimGrayaa, thanks for your understanding. We’ve updated the documentation and contributors should now create PRs to the `develop` branch.

Could you please change the base branch of your PR or open a new PR targeting `develop`?

**Comment by SalimGrayaa** [comment]
Hi @minh362000,

I rebased the branch onto `develop`, and I think everything is fine now. Could you please update the PR target from `main` to `develop`? I don’t seem to have permission to do it myself.

Or would you prefer that I create a new PR targeting `develop` instead?

**Comment by minh362000** [comment]
Hi @SalimGrayaa, I’ve already updated the PR target branch from main to develop, but the CI lint check is still failing.

Could you please run:
uv run black .
and
uv run ruff check --fix .
and push the updated changes?

**Comment by SalimGrayaa** [comment]
Hi @minh362000,
I’ve made the changes you requested. The lint checks should now pass.

File: packages/qdrant-loader/src/qdrant_loader/core/project_manager.py [hunk 1]
```diff
@@ -193,10 +194,24 @@ async def _ensure_project_in_database(
...
+        except IntegrityError as e:
+            # The unique-constraint error can be triggered by autoflush during
+            # session.execute() inside _update_project_sources, before commit.
+            await session.rollback()
+            err = str(e)
+            if "projects.collection_name" in err or "uix_project_collection" in err:
+                raise ValueError(
+                    "State DB schema is outdated and still enforces unique projects.collection_name. "
+                    "Reset state DB once and retry: `qdrant-loader init --workspace . --force`."
+                ) from e
```
[the remark this answers is not in the document]
Comment (minh362000) [reply packages/qdrant-loader/src/qdrant_loader/core/project_manager.py:210]: @codeRabbit Since it's not relevant to this PR, we will fix it later in another PR

File: packages/qdrant-loader/src/qdrant_loader/core/state/models.py [hunk 2]
```diff
@@ -70,10 +70,7 @@ class Project(Base):
         "DocumentStateRecord", back_populates="project", cascade="all, delete-orphan"
     )
 
-    __table_args__ = (
-        UniqueConstraint("collection_name", name="uix_project_collection"),
-        Index("ix_project_display_name", "display_name"),
-    )
+    __table_args__ = (Index("ix_project_display_name", "display_name"),)
```
[the remark this answers is not in the document]
Comment (minh362000) [reply packages/qdrant-loader/src/qdrant_loader/core/state/models.py:73]: @CodeRabbit Since it's not relevant to this PR, we will fix it later in another PR

File: uv.lock [hunk 3]
```diff
@@ -3613,7 +3613,7 @@ requires-dist = [
     { name = "httpx", specifier = ">=0.24.0" },
     { name = "langchain", specifier = ">=0.3.0" },
     { name = "langchain-community", specifier = ">=0.0.38" },
-    { name = "langchain-core", specifier = ">=1.3.3" },
+    { name = "langchain-core", specifier = ">=0.3.0" },
```
Comment (minh362000) [inline uv.lock]: Hi @SalimGrayaa, after @NguyenVu11796  and I checked the changes here, it's best if you keep the bumped versions because we have Dependabot, which checks and automatically bumps versions depending on the security concerns, such as leakage or vulnerabilities in the current version.

Comment (SalimGrayaa) [reply uv.lock]: Hi @minh362000,
I removed the changes made to the uv.lock file.

File: packages/qdrant-loader/src/qdrant_loader/connectors/jira/connector.py [hunk 4]
```diff
@@ -468,7 +474,53 @@ async def get_documents(self) -> list[Document]:
...
+                        }
+                        for att in issue.attachments
+                    ]
+                    if issue.attachments
+                    else []
+                ),
+            }
+            if self.config.extra_fields:
+                for field in self.config.extra_fields:
+                    metadata[field.name] = getattr(issue, field.name)
```
[the remark this answers is not in the document]
Comment (NguyenVu11796) [reply packages/qdrant-loader/src/qdrant_loader/connectors/jira/connector.py:523]: @cursor this is already fixed. The original bug was real: `cloud_connector.py` was calling `self._parse_issue(issue)` without passing `self.config.extra_fields`, so Cloud users would silently get no extra field extraction; that said, I think commit 6b28acb3 fixed it — the current code at line 108 is:


``parsed_issue = self._parse_issue(issue, self.config.extra_fields)``

File: packages/qdrant-loader/src/qdrant_loader/connectors/jira/mappers.py [hunk 5]
```diff
@@ -264,6 +291,16 @@ def _require_dict_with_key(
...
     )
+    if extra_fields:
+        for field in extra_fields:
+            value = _extract_extra_field_value(
+                fields,
+                field.param_name,
+                field.field_type,
+                field.attr_name,
+            )
+            setattr(jira_issue, field.name, value)
```
[the remark this answers is not in the document]
Comment (NguyenVu11796) [reply packages/qdrant-loader/src/qdrant_loader/connectors/jira/mappers.py:302]: @cursor This is also already fixed, and well I might add. The original concern was valid: `setattr(jira_issue, field.name, value)` with no guard would let a user name an extra field "id" or "key" and corrupt the document. The fix in commit 48a9973d is solid:

A `RESERVED_NAMES` set was added to config.py (all 19 built-in JiraIssue field names)
A` validate_reserved_name` `@model_validator` on `JiraExtraField` rejects any name in that set at config load time with a clear error message
The whitespace/empty validation from the CodeRabbit comment (`normalize_strings` validator) was added in the same pass
The defense is at the right layer — fail fast at config validation rather than at `setattr` time. No mapper-level guard needed since the config validator already blocks the bad input before it reaches the mapper.
